### PR TITLE
WIP: Merge cpp-project-template

### PR DESCRIPTION
**Pull request by MikaelSmith**
Merge the cpp-project-template to Leatherman, so we can pull in CI changes from it.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
TODO: fix coveralls in cpp-project-template.
TODO: fix Windows install in Leatherman.

**Comment by MikaelSmith** [comment]
Waiting on merge of https://github.com/puppetlabs/cpp-project-template/pull/20 and https://github.com/puppetlabs/leatherman/pull/108.

File: nowide/CMakeLists.txt [hunk 1]
```diff
@@ -12,18 +12,19 @@ if (BUILDING_LEATHERMAN)
     set(BOOST_NOWIDE_SKIP_INSTALL ON CACHE BOOL "Disable installing Boost.Nowide")
 
     # have to specify a bindir since this isn't actually a subdirectory of us
-    add_subdirectory("../vendor/boost-nowide" "../vendor/boost-nowide")
+    set(srcdir "../vendor/boost-nowide")
+    set(bindir "${CMAKE_CURRENT_BINARY_DIR}/${srcdir}")
```
Comment (branan) [inline nowide/CMakeLists.txt]: If we're making bindir based on `CMAKE_CURRENT_BINARY_DIR` we should probably go all in and make srcdir depend on `CMAKE_CURRENT_SOURCE_DIR`

Comment (branan) [inline nowide/CMakeLists.txt]: I guess that tweak would be in #108, now that I've worked my way up the PR stack

Comment (MikaelSmith) [inline nowide/CMakeLists.txt]: Yeah, I can do that there.

**Comment by MikaelSmith** [comment]
@adrienthebo and I debated this a little in https://github.com/puppetlabs/cppbestpractices/issues/17. I think I agree with him; we should be moving towards reducing churn in AppVeyor/Travis commits, and it's only a few files we can manually move over (or come up with a better way to just update those files).

**Pull request closed without merging**